### PR TITLE
Test automation starter

### PR DESCRIPTION
**Pull request by dkaravias**
Adding the whole app to master.
Worked with Sven.
The next PRs will be smaller and will reflect the work done for the [Test automation guidelines](https://novoda.atlassian.net/browse/PT-549)

### CODE REVIEW

File: AndroidTestAutomationStarter/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+## Overview
+
+A sample retail store app that showcases current testing methodologies for Android. It comes with a simple Heroku server that can be run locally.
+
+The aim is to provide a hands-on introduction to those new to test engineering and a challenge for those who want to further improve the current implementation.
+
+
+## Getting started
+
+Before launching the app you'll need to start the server. Make sure you have Ruby installed by running `ruby -v` from the terminal. If it's not installed run `brew install ruby`
```
[the remark this answers is not in the document]
Comment (dkaravias) [reply AndroidTestAutomationStarter/README.md]: It's the next line, where it says `Navigate to the simplesinatraapi directory in this project and run ruby server.rb` . The directory is included in the project.

**Comment by dkaravias** [comment]
@bartzet I formatted the whole project with https://github.com/shyiko/ktlint  
All Kotlin files should be consistent now.

File: AndroidTestAutomationStarter/app/src/androidTest/java/com/novoda/androidstoreexample/tests/EspressoJavaTestExample.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+import static android.support.test.espresso.Espresso.onView;
+import static android.support.test.espresso.action.ViewActions.click;
+import static android.support.test.espresso.assertion.ViewAssertions.matches;
+import static android.support.test.espresso.contrib.RecyclerViewActions.actionOnHolderItem;
+import static android.support.test.espresso.contrib.RecyclerViewActions.scrollToHolder;
+import static android.support.test.espresso.matcher.ViewMatchers.isDisplayed;
+import static android.support.test.espresso.matcher.ViewMatchers.withId;
+
+@RunWith(AndroidJUnit4.class)
```
[the remark this answers is not in the document]
Comment (dkaravias) [reply AndroidTestAutomationStarter/app/src/androidTest/java/com/novoda/androidstoreexample/tests/EspressoJavaTestExample.java]: Nice, tbh the whole test is redundant by now but we can leave it to have a Java/Kotlin side by side example (also demonstrating the difference between using or not using page objects)

File: AndroidTestAutomationStarter/app/src/androidTest/java/com/novoda/androidstoreexample/tests/EspressoUserFlowTests.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+    @get:Rule
+    var activityRule: ActivityTestRule<MainActivity> = activityTestRule
+
+    @Test
+    fun navigateToProductDetailsTest() {
+
+        mainActivityPageObject.navigateToProductList()
+
+        productListPageObject.navigateToProductDetails()
```
[the remark this answers is not in the document]
Comment (dkaravias) [reply AndroidTestAutomationStarter/app/src/androidTest/java/com/novoda/androidstoreexample/tests/EspressoUserFlowTests.kt]: These will be rewritten in https://novoda.atlassian.net/browse/PT-789

File: AndroidTestAutomationStarter/app/src/androidTest/java/com/novoda/androidstoreexample/tests/EspressoJavaTestExample.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+public class EspressoJavaTestExample {
+
+    @Rule
+    public ActivityTestRule<MainActivity> activityTestRule = new ActivityTestRule<MainActivity>(MainActivity.class);
+
+    @Test
+    public void firstNavigationTest() {
```
[the remark this answers is not in the document]
Comment (dkaravias) [reply AndroidTestAutomationStarter/app/src/androidTest/java/com/novoda/androidstoreexample/tests/EspressoJavaTestExample.java]: Just replied in the previous comment before seeing this one 😅

File: AndroidTestAutomationStarter/app/src/androidTest/java/com/novoda/androidstoreexample/tests/HeaderTests.kt [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class HeaderTests {
+    private val activityTestRule = ActivityTestRule<ProductListActivity>(ProductListActivity::class.java)
+
+    @Test
+    fun actionBarIsDisplayed() {
+        val headerPageObject = HeaderPageObject()
+        headerPageObject.verifyHeader()
```
[the remark this answers is not in the document]
Comment (dkaravias) [reply AndroidTestAutomationStarter/app/src/androidTest/java/com/novoda/androidstoreexample/tests/HeaderTests.kt]: It just checks if action_bar_container is displayed. Personally I'd be happy to get rid of it.

File: AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/activities/BaseActivity.kt [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+    abstract fun injectDependencies(appComponent: AppComponent)
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(getActivityLayout())
+        injectDependencies(App.component)
+    }
+
+    override fun showMessage(message: String) {
```
[the remark this answers is not in the document]
Comment (dkaravias) [reply AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/activities/BaseActivity.kt]: Not sure but when I remove it the app fails to compile (probably a different root cause but I'll look into it after merging)

File: AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/activities/BasketActivity.kt [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+    override fun showProgress() {
+    }
+
+    override fun hideProgress() {
+    }
+
+    override fun showBasketItems(orders: List<Order>) {
+        basketList.layoutManager = LinearLayoutManager(this)
+        basketAdapter = BasketAdapter(this, orders) {
```
[the remark this answers is not in the document]
Comment (dkaravias) [reply AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/activities/BasketActivity.kt]: Not sure, the basket is still WIP so I'd rather check with @thywen first

File: AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/dagger/App.kt [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+import android.app.Application
+import com.novoda.androidstoreexample.dagger.component.AppComponent
+import com.novoda.androidstoreexample.dagger.component.DaggerAppComponent
+
+class App : Application() {
+
+    companion object {
+        @JvmStatic
+        lateinit var instance: App
```
[the remark this answers is not in the document]
Comment (dkaravias) [reply AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/dagger/App.kt]: Also not sure but the app doesn't build without it

File: AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/dagger/module/AppModule.kt [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+import dagger.Provides
+import com.novoda.androidstoreexample.dagger.App
+import javax.inject.Singleton
+
+@Module
+class AppModule {
+
+    @Provides
+    @Singleton
+    fun provideContext(): Context = App.instance
```
[the remark this answers is not in the document]
Comment (dkaravias) [reply AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/dagger/module/AppModule.kt]: Hm would you be ok to pair on this when you're back? I'm not familiar enough with Dagger

File: AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/dagger/module/ClientModule.kt [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+import okhttp3.OkHttpClient
+import java.util.concurrent.TimeUnit
+import javax.inject.Singleton
+
+@Module
+class ClientModule {
+
+    @Provides
+    @Singleton
+    fun provideClient(networkTimeout: Long): OkHttpClient {
```
[the remark this answers is not in the document]
Comment (dkaravias) [reply AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/dagger/module/ClientModule.kt]: Same as previous comment, I'd rather pair on it

File: AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/dagger/module/ClientModule.kt [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+        val okHttpClientBuilder: OkHttpClient.Builder = OkHttpClient.Builder()
```
[the remark this answers is not in the document]
Comment (dkaravias) [reply AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/dagger/module/ClientModule.kt]: 💯

File: AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/mvp/interactor/impl/CategoryListInteractorImpl.kt [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+import com.novoda.androidstoreexample.mvp.interactor.CategoryListInteractor
+import com.novoda.androidstoreexample.mvp.listener.CategoryListListener
+import com.novoda.androidstoreexample.services.ShopService
+import retrofit2.Call
+import retrofit2.Callback
+import retrofit2.Response
+import retrofit2.Retrofit
+import javax.inject.Inject
+
+class CategoryListInteractorImpl : CategoryListInteractor {
```
[the remark this answers is not in the document]
Comment (dkaravias) [reply AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/mvp/interactor/impl/CategoryListInteractorImpl.kt]: Not sure what you mean, again happy to catch up when you're back

File: AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/mvp/presenter/impl/ProductDetailPresenterImpl.kt [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+import com.novoda.androidstoreexample.services.BasketService
+import com.novoda.androidstoreexample.services.ProductDetailsResponse
+import javax.inject.Inject
+
+class ProductDetailPresenterImpl : ProductDetailPresenter {
+    private val productDetailView: ProductDetailView
+    private val productDetailsInteractor: ProductDetailsInteractor
+    private val basketService: BasketService
+
+    private lateinit var product: Product
```
[the remark this answers is not in the document]
Comment (dkaravias) [reply AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/mvp/presenter/impl/ProductDetailPresenterImpl.kt]: I don't think we want to add without a product!

File: AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/services/CategoryResponse.kt [hunk 14]
```diff
@@ -0,0 +1,21 @@
+package com.novoda.androidstoreexample.services
+
+import com.novoda.androidstoreexample.models.Category
+import java.util.*
+
+data class CategoryResponse(val categories: Array<Category>) {
+    override fun equals(other: Any?): Boolean {
```
[the remark this answers is not in the document]
Comment (dkaravias) [reply AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/services/CategoryResponse.kt]: I'd rather check with @thywen who wrote this

File: AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/services/ProductResponse.kt [hunk 15]
```diff
@@ -0,0 +1,21 @@
+package com.novoda.androidstoreexample.services
+
+import com.novoda.androidstoreexample.models.Product
+import java.util.*
+
+data class ProductResponse(val products: Array<Product>) {
+    override fun equals(other: Any?): Boolean {
```
[the remark this answers is not in the document]
Comment (dkaravias) [reply AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/services/ProductResponse.kt]: same as above

File: AndroidTestAutomationStarter/simplesinatraapi/item_repository.rb [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+
+  def items_from_category(category)
+    @items[category]
+  end
+
+  def get_random_description(number_of_sentences=2)
+    Faker::Hipster.sentences(number_of_sentences).join(' ')
+  end
+
+  def get_item_for_id(id)
```
[the remark this answers is not in the document]
Comment (dkaravias) [reply AndroidTestAutomationStarter/simplesinatraapi/item_repository.rb]: Again, care to explain when you have some time?

File: AndroidTestAutomationStarter/app/src/androidTest/java/com/novoda/androidstoreexample/tests/HeaderTests.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (fourlastor) [reply AndroidTestAutomationStarter/app/src/androidTest/java/com/novoda/androidstoreexample/tests/HeaderTests.kt]: same, or at least renaming it `assertActionBarIsDisplayed`

File: AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/dagger/module/AppModule.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (fourlastor) [reply AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/dagger/module/AppModule.kt]: yep, no problem

File: AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/mvp/interactor/impl/CategoryListInteractorImpl.kt [hunk 17]
```diff
@@ -1,32 +1,24 @@
...
-        this.retrofit = retrofit
-        this.apiService = apiService
-    }
+    private var call: Call<CategoryResponse>? = null
 
     override fun loadCategoryList(categoryListListener: CategoryListListener) {
         call = apiService.getCategories()
-        call.enqueue(object : Callback<CategoryResponse> {
+        call?.enqueue(object : Callback<CategoryResponse> {
             override fun onResponse(call: Call<CategoryResponse>?, response: Response<CategoryResponse>?) {
```
Comment (fourlastor) [inline AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/mvp/interactor/impl/CategoryListInteractorImpl.kt]: I think you can remove the nullables also here

File: AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/mvp/interactor/impl/ProductListInteractorImpl.kt [hunk 18]
```diff
@@ -1,45 +1,38 @@
...
-    override fun cancel() {
-        call.cancel()
-    }
+    private var call: Call<ProductResponse>? = null
 
     override fun loadProductList(produListListener: ProductListListener, category: Int) {
         call = apiService.getProductsFromCategory(category)
-        call.enqueue(object : Callback<ProductResponse> {
+        call?.enqueue(object : Callback<ProductResponse> {
             override fun onFailure(call: Call<ProductResponse>?, t: Throwable?) {
```
Comment (fourlastor) [inline AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/mvp/interactor/impl/ProductListInteractorImpl.kt]: and here (removing the nullables)

File: AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/activities/MainActivity.kt [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+    override fun getActivityLayout(): Int {
+        return R.layout.activity_main
+    }
+
+    override fun injectDependencies(appComponent: AppComponent) {
+        appComponent.injectCategory(CategoryListModule(this)).inject(this)
+    }
+
+    override fun onItemClicked(type: Int) {
+        intent = Intent(this, ProductListActivity::class.java).apply { putExtra(CATEGORY_ID_EXTRA, type) }
```
Comment (fourlastor) [inline AndroidTestAutomationStarter/app/src/main/java/com/novoda/androidstoreexample/activities/MainActivity.kt]: this is setting the activity intent ⚠️ be sure you use `val`

**Comment by jgtaylor123** [comment]
Gave it a walk through, tests are running.  Looks good.

**Pull request merged**